site/guides/effective-go-krm-function.md

```diff
@@ -0,0 +1,212 @@
+# Effective GO KRM functions
+
+This guide gives tips to effectively write a KRM function. 
+
+This guide is for advanced kpt function users who find the [catalog.kpt.dev] cannot fulfil their needs 
+and want to design their own KRM functions. 
+
+Suggest reading [Developing in Go] first
+
+## Prerequisites
+
+- [Install kpt]
+- [Install Docker]
+
+## Setup 
+
+<!--- TODO: Use scaffolding to generate the get-started package --->
+
+We start from a "get-started-runner" package which contains a `main.go` file with some scaffolding code.
+
+
+```shell
+# Set your KRM function name.
+export FUNCTION_NAME=<YOUR FUNCTION NAME>
+export GOPATH=$(go env GOPATH)
+export FUNCTION_PATH=github.com/<YOUR USERNAME>
+
+# Create and direct to your Go working directory
+mkdir -p $GOPATH/src/${FUNCTION_PATH} && cd $GOPATH/src/${FUNCTION_PATH}
+
+# Get the "get-started" package.
+kpt pkg get https://github.com/GoogleContainerTools/kpt-functions-sdk.git/go/get-started-runner@master ${FUNCTION_NAME}
+
+cd ${FUNCTION_NAME}
+
+# Initialize Go module and install the kpt KRM function SDK.
+go mod init && go mod tidy -compat=1.17
+```
+
+## Write your KRM function code in Go 
+
+In the main.go, you should have
+```go
+var _ fn.Runner = &FunctionX{}
+
+type FunctionX struct {
+    // TODO: Modify with your expected function config.
+    FnConfigBool bool
+    FnConfigInt  int
+    FnConfigFoo string
+}
+
+func (r *FunctionX) Run(ctx *fn.Context, functionConfig *fn.KubeObject, items fn.KubeObjects) {
+      // TODO: Add your KRM resource mutate or validate logic.
+}
+
+func main() {
+    if err := fn.AsMain(&FunctionX{}); err != nil {
+        os.Exit(1)
+    }
+}
+```
+`FunctionX` implements the [`Runner`] interface that can process the input KRM resources as [`fn.ResourceList`], it initializes `fn.KubeObject` to hold the KRM resources,
+so that you can use [`fn.KubeObject` and `fn.SubObject`] methods directly. After `Run`, it will convert the modified `fn.KubeObjects` to KRM resources.
```

Comment:
Instead of "it initializes `fn.KubeObject` to hold the KRM resources, so that you can use [`fn.KubeObject` and `fn.SubObject`] methods directly", I would just say that something like "the objects passed to the function are available in items.  These objects can be modified in place and are used as the output of the function".